### PR TITLE
Groups assign tasks

### PR DESCRIPTION
**Pull request by TheHollidayInn**
Fixes put_issue_url_here
### Changes

This depends on #7785.

This adds the ability for a guild leader to assign and unassign users.

This will also create a separate task for assigned users and sync the task when it is updated or deleted.

---

UUID:

### CODE REVIEW

File: test/api/v3/integration/dataexport/GET-export_history.csv.test.js [hunk 1]
```diff
@@ -39,12 +39,13 @@ describe('GET /export/history.csv', () => {
...
     let res = await user.get('/export/history.csv');
     let splitRes = res.split('\n');
+
     expect(splitRes[0]).to.equal('Task Name,Task ID,Task Type,Date,Value');
     expect(splitRes[1]).to.equal(`habit 1,${tasks[0]._id},habit,${moment(tasks[0].history[0].date).format('YYYY-MM-DD HH:mm:ss')},${tasks[0].history[0].value}`);
     expect(splitRes[2]).to.equal(`habit 1,${tasks[0]._id},habit,${moment(tasks[0].history[1].date).format('YYYY-MM-DD HH:mm:ss')},${tasks[0].history[1].value}`);
-    expect(splitRes[3]).to.equal(`daily 1,${tasks[1]._id},daily,${moment(tasks[1].history[0].date).format('YYYY-MM-DD HH:mm:ss')},${tasks[1].history[0].value}`);
-    expect(splitRes[4]).to.equal(`habit 2,${tasks[2]._id},habit,${moment(tasks[2].history[0].date).format('YYYY-MM-DD HH:mm:ss')},${tasks[2].history[0].value}`);
-    expect(splitRes[5]).to.equal(`habit 2,${tasks[2]._id},habit,${moment(tasks[2].history[1].date).format('YYYY-MM-DD HH:mm:ss')},${tasks[2].history[1].value}`);
+    expect(splitRes[3]).to.equal(`habit 2,${tasks[2]._id},habit,${moment(tasks[2].history[0].date).format('YYYY-MM-DD HH:mm:ss')},${tasks[2].history[0].value}`);
```
Comment (paglias) [inline test/api/v3/integration/dataexport/GET-export_history.csv.test.js:46]: why you had to change this test?

File: website/server/controllers/api-v3/tasks.js [hunk 2]
```diff
@@ -341,6 +234,14 @@ api.updateTask = {
     // see https://github.com/Automattic/mongoose/issues/2749
 
     let savedTask = await task.save();
+
+    if (task.group.id && task.assignedUsers.length > 0) {
+      let group = await Group.getGroup({user, groupId: task.group.id, populateLeader: false});
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: you should limit the fields fetched here or at least exclude the chat because it's very large

File: website/server/controllers/api-v3/tasks.js [hunk 3]
```diff
@@ -951,6 +852,14 @@ api.deleteTask = {
       throw new NotAuthorized(res.t('cantDeleteChallengeTasks'));
     }
 
+    if (task.group.id) {
+      //  @TODO: Abstract this access snippet
+      let group = await Group.getGroup({user, groupId: task.group.id, populateLeader: false});
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: same here, limit the fetched fields

File: website/server/controllers/api-v3/tasks.js [hunk 4]
```diff
@@ -951,6 +852,14 @@ api.deleteTask = {
...
+      if (!group) throw new NotFound(res.t('groupNotFound'));
+      if (group.leader !== user._id) throw new NotAuthorized(res.t('onlyGroupLeaderCanEditTasks'));
+      group.removeTask(task);
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: await?

File: website/server/controllers/api-v3/tasks/groups.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    let assignedUser = await User.findById(assignedUserId);
+
+    let taskId = req.params.taskId;
+    let task = await Tasks.Task.findByIdOrAlias(taskId, user._id);
+
+    if (!task) {
+      throw new NotFound(res.t('taskNotFound'));
+    }
+
+    if (!task.group || !task.group.id) {
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: checking for task.group should not be necessary as it'll always exist, even if empty

File: website/server/controllers/api-v3/tasks/groups.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+      throw new NotAuthorized(res.t('onlyGroupTasksCanBeAssigned'));
+    }
+
+    let group = await Group.getGroup({user, groupId: task.group.id, populateLeader: false});
+    if (!group) throw new NotFound(res.t('groupNotFound'));
+
+    if (group.leader !== user._id) throw new NotAuthorized(res.t('onlyGroupLeaderCanEditTasks'));
+
+    group.syncTask(task, assignedUser);
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: await?

File: website/server/controllers/api-v3/tasks/groups.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    let assignedUser = await User.findById(assignedUserId);
+
+    let taskId = req.params.taskId;
+    let task = await Tasks.Task.findByIdOrAlias(taskId, user._id);
+
+    if (!task) {
+      throw new NotFound(res.t('taskNotFound'));
+    }
+
+    if (!task.group || !task.group.id) {
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: same as above about checking for task.group

File: website/server/controllers/api-v3/tasks/groups.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+      throw new NotAuthorized(res.t('onlyGroupTasksCanBeAssigned'));
+    }
+
+    let group = await Group.getGroup({user, groupId: task.group.id, populateLeader: false});
+    if (!group) throw new NotFound(res.t('groupNotFound'));
+
+    if (group.leader !== user._id) throw new NotAuthorized(res.t('onlyGroupLeaderCanEditTasks'));
+
+    group.unlinkTask(task, assignedUser);
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: await?

File: website/server/controllers/api-v3/tasks/groups.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+    res.respond(201, task);
+
+    return null;
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: this is not necessary

File: website/server/controllers/api-v3/tasks/groups.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    let group = await Group.getGroup({user, groupId: task.group.id, populateLeader: false});
+    if (!group) throw new NotFound(res.t('groupNotFound'));
+
+    if (group.leader !== user._id) throw new NotAuthorized(res.t('onlyGroupLeaderCanEditTasks'));
+
+    group.syncTask(task, assignedUser);
+
+    res.respond(201, task);
+
+    return null;
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: not necessary

File: website/server/controllers/api-v3/tasks/groups.js [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    let group = await Group.getGroup({user, groupId: req.params.groupId, populateLeader: false});
+    if (!group) throw new NotFound(res.t('groupNotFound'));
+
+    if (group.leader !== user._id) throw new NotAuthorized(res.t('onlyGroupLeaderCanEditTasks'));
+
+    let tasks = await createTasks(req, res, {user, undefined, group});
+
+    res.respond(201, tasks.length === 1 ? tasks[0] : tasks);
+
+    return null;
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: not necessary

File: website/server/libs/api-v3/taskManager.js [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+  // Compares the short names in tasks against
+  // a Set, where values cannot repeat. If the
+  // lengths are different, some name was duplicated
+  if (aliases.length !== [...new Set(aliases)].length) {
+    throw new BadRequest(res.t('taskAliasAlreadyUsed'));
+  }
+
+  await Bluebird.map(tasksWithAliases, (task) => {
+    return task.validate();
```
Comment (paglias) [inline website/server/libs/api-v3/taskManager.js]: This is not strictly related to this PR but since we don't have async mongoose validators (as far as i remember) we could change this to use `task.validateSync()` which should be considerably faster as it's only sync code

File: website/server/libs/api-v3/taskManager.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    return tasks;
+  }
+}
+
+// Takes a Task document and return a plain object of attributes that can be synced to the user
+
+export function syncableAttrs (task) {
+  let t = task.toObject(); // lodash doesn't seem to like _.omit on Document
+  // only sync/compare important attrs
+  let omitAttrs = ['_id', 'userId', 'challenge', 'history', 'tags', 'completed', 'streak', 'notes', 'updatedAt'];
```
Comment (paglias) [inline website/server/libs/api-v3/taskManager.js]: this should probably include `group` too now

File: website/server/models/group.js [hunk 14]
```diff
@@ -748,6 +756,113 @@ schema.methods.leave = async function leaveGroup (user, keep = 'keep-all') {
...
+    _.merge(matchingTask, syncableAttrs(taskToSync));
+    // Make sure the task is in user.tasksOrder
+    let orderList = user.tasksOrder[`${taskToSync.type}s`];
+    if (orderList.indexOf(matchingTask._id) === -1 && (matchingTask.type !== 'todo' || !matchingTask.completed)) orderList.push(matchingTask._id);
+  }
+
+  if (!matchingTask.notes) matchingTask.notes = taskToSync.notes; // don't override the notes, but provide it if not provided
+  if (matchingTask.tags.indexOf(group._id) === -1) matchingTask.tags.push(group._id); // add tag if missing
+
+  toSave.push(matchingTask.save());
```
Comment (paglias) [inline website/server/models/group.js]: toSave.push(matchingTask.save(), taskToSync.save(), ...)

File: website/server/models/group.js [hunk 15]
```diff
@@ -748,6 +756,113 @@ schema.methods.leave = async function leaveGroup (user, keep = 'keep-all') {
...
+    linkedTaskId: unlinkingTask._id,
+    userId: user._id,
+  };
+
+  let assignedUserIndex = unlinkingTask.assignedUsers.indexOf(user._id);
+  unlinkingTask.assignedUsers.splice(assignedUserIndex, 1);
+
+  if (keep === 'keep-all') {
+    await Tasks.Task.update(findQuery, {
+      $set: {'group.id': undefined},
```
Comment (paglias) [inline website/server/models/group.js]: $unset group.id?

File: website/server/models/task.js [hunk 16]
```diff
@@ -57,6 +57,8 @@ export let TaskSchema = new Schema({
   },
   attribute: {type: String, default: 'str', enum: ['str', 'con', 'int', 'per']},
   userId: {type: String, ref: 'User', validate: [validator.isUUID, 'Invalid uuid.']}, // When not set it belongs to a challenge
+  assignedUsers: [{type: String, ref: 'User', validate: [validator.isUUID, 'Invalid uuid.']}],
```
Comment (paglias) [inline website/server/models/task.js]: these two should be added to the list of not settable attributes

File: website/server/libs/taskManager.js [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+  });
+}
+
+
+/**
+ * Creates tasks for a user, challenge or group.
+ *
+ * @param  req  The Express req variable
+ * @param  res  The Express res variable
+ * @param  options  Should be a user, user and challenge to create a challenge task or user and group to create a group task
```
Comment (crookedneighbor) [inline website/server/libs/taskManager.js]: I like this, but let's be more specific:

``` js
* @param  options 
* @param  options.user  The user that these tasks belong to
* @param  options.challenge  The challenge that these tasks belong to
* @param  options.group  The group that these tasks belong to
```

File: website/server/libs/taskManager.js [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+  toSave = toSave.map(taskData => {
+    // Validate that task.type is valid
+    if (!taskData || Tasks.tasksTypes.indexOf(taskData.type) === -1) throw new BadRequest(res.t('invalidTaskType'));
+
+    let taskType = taskData.type;
+    let newTask = new Tasks[taskType](Tasks.Task.sanitize(taskData));
+
+    if (challenge) {
+      newTask.challenge.id = challenge.id;
+    } else {
```
Comment (crookedneighbor) [inline website/server/libs/taskManager.js:58]: Is this logic correct? Do you always needs to pass in a user when passing in a group?

Comment (paglias) [inline website/server/libs/taskManager.js:58]: Hmm yeah... if it works like challenges tasks do then it there should be no userId when it's the original group task

Comment (paglias) [inline website/server/libs/taskManager.js:58]: About this comment, there's this code https://github.com/HabitRPG/habitrpg/pull/7887/files#diff-7e3afbc9a950c0819d20e0d1c440ee4aR209 that loads the challenge if there's no `task.userId` it'll load the challenge automatically which should not happen if the task is linked to a group instead

File: website/server/controllers/api-v3/tasks.js [hunk 19]
```diff
@@ -962,6 +851,14 @@ api.deleteTask = {
       throw new NotAuthorized(res.t('cantDeleteChallengeTasks'));
     }
 
+    if (task.group.id) {
+      //  @TODO: Abstract this access snippet
+      let group = await Group.getGroup({user, groupId: req.params.groupId, populateLeader: false, fields: '_id'});
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: populateLeader is false by default

File: test/api/v3/unit/libs/taskManager.js [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+    expect(task.down).to.equal(testHabit.down);
+    expect(task.createdAt).isNotEmtpy;
+    expect(task.challenge.id).to.equal(challenge._id);
+  });
+
+  it('returns syncable attibutes', async () => {
+    req.body = testHabit;
+    res.t = i18n.t;
+
+    let tasks = await createTasks(req, res, {user, challenge, undefined});
```
Comment (paglias) [inline test/api/v3/unit/libs/taskManager.js]: undefined

File: website/server/controllers/api-v3/tasks/groups.js [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+    if (reqValidationErrors) throw reqValidationErrors;
+
+    let user = res.locals.user;
+
+    let group = await Group.getGroup({user, groupId: req.params.groupId, populateLeader: false, fields: '_id'});
+    if (!group) throw new NotFound(res.t('groupNotFound'));
+
+    if (group.leader !== user._id) throw new NotAuthorized(res.t('onlyGroupLeaderCanEditTasks'));
+
+    let tasks = await createTasks(req, res, {user, undefined, group});
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: undefined

File: website/server/controllers/api-v3/tasks/groups.js [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+  middlewares: [authWithHeaders()],
+  async handler (req, res) {
+    req.checkParams('groupId', res.t('groupIdRequired')).notEmpty().isUUID();
+
+    let reqValidationErrors = req.validationErrors();
+    if (reqValidationErrors) throw reqValidationErrors;
+
+    let user = res.locals.user;
+
+    let group = await Group.getGroup({user, groupId: req.params.groupId, populateLeader: false, fields: '_id'});
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: populateLeader: false by default

File: website/server/controllers/api-v3/tasks/groups.js [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+ *
+ * @apiSuccess {Array} data An array of tasks
+ */
+api.getGroupTasks = {
+  method: 'GET',
+  url: '/tasks/group/:groupId',
+  middlewares: [authWithHeaders()],
+  async handler (req, res) {
+    req.checkParams('groupId', res.t('groupIdRequired')).notEmpty().isUUID();
+    let types = Tasks.tasksTypes.map(type => `${type}s`);
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: I think this can go outside the controller so it's not calculated anytime

File: website/server/controllers/api-v3/tasks/groups.js [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+    req.checkQuery('type', res.t('invalidTaskType')).optional().isIn(types);
+
+    let validationErrors = req.validationErrors();
+    if (validationErrors) throw validationErrors;
+
+    let user = res.locals.user;
+
+    let group = await Group.getGroup({user, groupId: req.params.groupId, populateLeader: false, fields: '_id'});
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: populateLeader: false by default

File: website/server/controllers/api-v3/tasks/groups.js [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+
+    if (!task) {
+      throw new NotFound(res.t('taskNotFound'));
+    }
+
+    if (!task.group.id) {
+      throw new NotAuthorized(res.t('onlyGroupTasksCanBeAssigned'));
+    }
+
+    let group = await Group.getGroup({user, groupId: req.params.groupId, populateLeader: false, fields: '_id'});
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: populateLeader: false by default

File: website/server/controllers/api-v3/tasks/groups.js [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+    if (!group) throw new NotFound(res.t('groupNotFound'));
+
+    if (group.leader !== user._id) throw new NotAuthorized(res.t('onlyGroupLeaderCanEditTasks'));
+
+    await group.syncTask(task, assignedUser);
+
+    res.respond(201, task);
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: 200 not 201

File: website/server/controllers/api-v3/tasks/groups.js [hunk 27]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    let group = await Group.getGroup({user, groupId: req.params.groupId, populateLeader: false, fields: '_id'});
+    if (!group) throw new NotFound(res.t('groupNotFound'));
+
+    if (group.leader !== user._id) throw new NotAuthorized(res.t('onlyGroupLeaderCanEditTasks'));
+
+    await group.unlinkTask(task, assignedUser);
+
+    res.respond(201, task);
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: 200

File: website/server/controllers/api-v3/tasks/groups.js [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+
+    if (!task) {
+      throw new NotFound(res.t('taskNotFound'));
+    }
+
+    if (!task.group.id) {
+      throw new NotAuthorized(res.t('onlyGroupTasksCanBeAssigned'));
+    }
+
+    let group = await Group.getGroup({user, groupId: req.params.groupId, populateLeader: false, fields: '_id'});
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: populateLeader: false by default

File: website/server/libs/taskManager.js (hunk 17 above)
Comment (paglias) [inline website/server/libs/taskManager.js]: If we decide to use this doc style we should try to be consistent on how and where we use it

File: website/server/libs/taskManager.js [hunk 29]
```diff
@@ -0,0 +1,170 @@
...
+  });
+
+  // tasks with aliases need to be validated asyncronously
+  await _validateTaskAlias(toSave, res);
+
+  toSave = toSave.map(task => task.save({ // If all tasks are valid (this is why it's not in the previous .map()), save everything, withough running validation again
+    validateBeforeSave: false,
+  }));
+
+  toSave.unshift((challenge || user).save());
```
Comment (paglias) [inline website/server/libs/taskManager.js]: group is missing?

File: website/server/libs/taskManager.js [hunk 30]
```diff
@@ -0,0 +1,170 @@
...
+    challenge,
+    group,
+  } = options;
+
+  let query = {userId: user._id};
+
+  if (challenge) {
+    query =  {'challenge.id': challenge.id, userId: {$exists: false}};
+  } else if (group) {
+    query =  {'group.id': group._id};
```
Comment (paglias) [inline website/server/libs/taskManager.js]: userId: {$exists: false}

File: website/server/libs/taskManager.js [hunk 31]
```diff
@@ -0,0 +1,170 @@
...
+      {type: 'todo', completed: false},
+      {type: {$in: ['habit', 'daily', 'reward']}},
+    ];
+  }
+
+  let tasks = await Tasks.Task.find(query).exec();
+
+  // Order tasks based on tasksOrder
+  if (type && type !== 'completedTodos' && type !== '_allCompletedTodos') {
+    let order = (challenge || user).tasksOrder[type];
```
Comment (paglias) [inline website/server/libs/taskManager.js]: group

File: website/server/libs/taskManager.js [hunk 32]
```diff
@@ -0,0 +1,170 @@
...
+    return tasks;
+  }
+}
+
+// Takes a Task document and return a plain object of attributes that can be synced to the user
+
+export function syncableAttrs (task) {
+  let t = task.toObject(); // lodash doesn't seem to like _.omit on Document
+  // only sync/compare important attrs
+  let omitAttrs = ['_id', 'userId', 'challenge', 'history', 'tags', 'completed', 'streak', 'notes', 'updatedAt', 'group', 'assignedUsers', 'linkedTaskId'];
```
Comment (paglias) [inline website/server/libs/taskManager.js]: linkedTaskId should be under task.group?

File: website/server/models/group.js [hunk 33]
```diff
@@ -760,6 +770,168 @@ schema.methods.leave = async function leaveGroup (user, keep = 'keep-all') {
   return await Bluebird.all(promises);
 };
 
+schema.methods.updateTask = async function updateTask (taskToSync) {
+  let group = this;
+  let listOfPromises = [];
+
+  let usersToSync = await User.find({_id: { $in: taskToSync.assignedUsers} });
+
+  usersToSync.forEach(function syncAssignedUser (user) {
```
Comment (paglias) [inline website/server/models/group.js]: usersToSync can be really heavy with many users so it should only be used when there's only one user involved. When you have to update the tasks for many users you should directly issue an update command to mongo like we do for challenges https://github.com/HabitRPG/habitrpg/blob/develop/website/server/models/challenge.js#L202

File: website/server/models/group.js [hunk 34]
```diff
@@ -760,6 +770,168 @@ schema.methods.leave = async function leaveGroup (user, keep = 'keep-all') {
...
+    linkedTaskId: unlinkingTask._id,
+    userId: user._id,
+  };
+
+  let assignedUserIndex = unlinkingTask.assignedUsers.indexOf(user._id);
+  unlinkingTask.assignedUsers.splice(assignedUserIndex, 1);
+
+  if (keep === 'keep-all') {
+    await Tasks.Task.update(findQuery, {
+      $unset: {'group.id': ''},
```
Comment (paglias) [inline website/server/models/group.js]: maybe this should set task.group to an empty object? for challenges we do `$set: {challenge: {}},`

File: website/server/models/group.js [hunk 35]
```diff
@@ -760,6 +770,168 @@ schema.methods.leave = async function leaveGroup (user, keep = 'keep-all') {
...
+    userId: {$exists: true},
+    'group.id': group.id,
+    linkedTaskId: task._id,
+  }, {
+    $set: {'group.broken': 'TASK_DELETED'},
+  }, {multi: true}).exec();
+};
+
+// API v2 compatibility methods
+schema.methods.getTransformedData = function getTransformedData (options) {
```
Comment (paglias) [inline website/server/models/group.js]: This is a leftover from when API v2 was still there, should be removed

File: website/server/models/task.js [hunk 36]
```diff
@@ -64,6 +66,11 @@ export let TaskSchema = new Schema({
     winner: String, // user.profile.name of the winner
   },
 
+  group: {
+    id: {type: String, ref: 'Group', validate: [validator.isUUID, 'Invalid uuid.']},
+    broken: {type: String, enum: ['GROUP_DELETED', 'TASK_DELETED', 'UNSUBSCRIBED', 'GROUP_TASK_NOT_FOUND']}, // CHALLENGE_TASK_NOT_FOUND comes from v3 migration
```
Comment (paglias) [inline website/server/models/task.js]: I think `GROUP_TASK_NOT_FOUND` doesn't really apply for groups but only for challenges

File: website/server/models/group.js (hunk 34 above)
Comment (TheHollidayInn) [inline website/server/models/group.js]: You told me to change this to $unset above.

File: website/server/controllers/api-v3/tasks/groups.js (hunk 27 above)
Comment (TheHollidayInn) [inline website/server/controllers/api-v3/tasks/groups.js]: Why is this 201 then: https://github.com/HabitRPG/habitrpg/blob/develop/website/server/controllers/api-v3/tasks.js#L92 Is 201 for content created?

File: website/server/models/group.js (hunk 34 above)
Comment (paglias) [inline website/server/models/group.js]: Yeah I should have suggested to use $set before, sorry. The $unset was only for group.id but it's better to reset the entire object

File: website/server/controllers/api-v3/tasks/groups.js (hunk 27 above)
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: yeah in theory 201 means "CREATED"

File: website/server/models/group.js (hunk 34 above)
Comment (TheHollidayInn) [inline website/server/models/group.js]: Yea, I just copied from the Challenge function before, no worries.

File: website/server/controllers/api-v3/tasks/groups.js (hunk 27 above)
Comment (TheHollidayInn) [inline website/server/controllers/api-v3/tasks/groups.js]: Okay, I'll keep that in mind.

File: website/server/libs/taskManager.js (hunk 32 above)
Comment (TheHollidayInn) [inline website/server/libs/taskManager.js]: I think it would be useful to have this for challenge and group - and in the future other items. Not sure if you agree.

File: website/server/models/task.js (hunk 36 above)
Comment (TheHollidayInn) [inline website/server/models/task.js]: Even though these are linked to group tasks?

File: website/server/libs/taskManager.js (hunk 32 above)
Comment (paglias) [inline website/server/libs/taskManager.js]: Hmm it could make sense but the problem is that the linked task for challenges it's stored under task.challenge.taskId and it would be difficult to migrate it now

File: website/server/models/task.js (hunk 36 above)
Comment (paglias) [inline website/server/models/task.js]: yeah because we don't don't really use CHALLENGE_TASK_NOT_FOUND but I messed up with the v3 migration so some task got unlinked even though it shouldn't have. But it's not really used in the code, it was just to identify what needed to be fixed from the migration

File: website/server/libs/taskManager.js (hunk 17 above)
Comment (crookedneighbor) [inline website/server/libs/taskManager.js]: Agreed.

File: website/server/libs/taskManager.js (hunk 32 above)
Comment (paglias) [inline website/server/libs/taskManager.js]: same thing for assignedUsers

File: website/server/libs/taskManager.js (hunk 29 above)
Comment (crookedneighbor) [inline website/server/libs/taskManager.js]: Instead of passing `(group || challenge || user)` everywhere, we could do

``` js
let owner = group || challenge || user;
```

File: website/server/libs/taskManager.js (hunk 18 above)
Comment (TheHollidayInn) [inline website/server/libs/taskManager.js:58]: I was thinking we don't depend on not having a task.userId. I think depending on the task.linkTaskId is more reliable and understandable. Challenge has a linkedId as well, so the idea of assuming a task is a challenge task if it doesn't have a user is implicit when the logic could be explicit.

File: website/server/libs/taskManager.js (hunk 32 above)
Comment (TheHollidayInn) [inline website/server/libs/taskManager.js]: Okay, but I think we should think more generally though. What if we want assigned users from challenges or something else in the future. The linkedTaskID is an example. We didn't generalize the challenge linked id, now we have two fields that are essentially the same.

**Comment by crookedneighbor** [comment]
Killed your build because it failed on the task manager tests. `owner.save` is not a function

**Comment by paglias** [comment]
Can we call task.group.linkedTaskId task.group.taskId so that it matches task.challenge.taskId?

**Comment by crookedneighbor** [comment]
I agree with @paglias

**Comment by TheHollidayInn** [comment]
That makes sense. I'll update that

File: website/server/libs/api-v3/taskManager.js (hunk 12 above)
Comment (TheHollidayInn) [inline website/server/libs/api-v3/taskManager.js]: I had to revert this change because it breaks https://github.com/HabitRPG/habitrpg/blob/develop/website/server/models/task.js#L164

**Comment by TheHollidayInn** [comment]
Updated.

**Comment by paglias** [comment]
I think it's ready to be merged. @crookedneighbor opinions?

**Comment by paglias** [comment]
Maybe if we merge it we should block the routes from being usable in production?

Also, there are currently no checks on whether the group is part of a group plan and thus can use the task assigning feature. Like a flag on the group model

**Comment by TheHollidayInn** [comment]
Updated with middleware to prevent use on the live site. The payments will be in a separate PR.

File: website/server/controllers/api-v3/tasks/groups.js [hunk 37]
```diff
@@ -1,5 +1,5 @@
 import { authWithHeaders } from '../../../middlewares/auth';
-import { ensureDevelpmentMode } from '../../../middlewares/ensureDevelpmentMode';
+import ensureDevelpmentMode from '../../../middlewares/ensureDevelpmentMode';
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js:2]: hmm we're missing an `o` in the file name. ensureDevelOpmentMode

Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks/groups.js:2]: Ah, yea. Looks like it was merged in like that. That was my bad. I can fix that Ina separate PR. Other files will be affected.

Comment (paglias) [reply website/server/controllers/api-v3/tasks/groups.js:2]: Sounds good

File: website/server/controllers/api-v3/tasks.js [hunk 38]
```diff
@@ -634,9 +634,18 @@ api.addTagToTask = {
     if (validationErrors) throw validationErrors;
 
     let taskId = req.params.taskId;
-    let task = await Tasks.Task.findByIdOrAlias(taskId, user._id, { userId: user._id });
+    let task = await Tasks.Task.findByIdOrAlias(taskId, user._id);
+
+    if (!task) {
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: `if (!task || task.userId !== user._id)` since the last `else if` can be added to the first one

File: website/server/controllers/api-v3/tasks.js [hunk 39]
```diff
@@ -674,9 +683,17 @@ api.removeTagFromTask = {
     if (validationErrors) throw validationErrors;
 
     let taskId = req.params.taskId;
-    let task = await Tasks.Task.findByIdOrAlias(taskId, user._id, { userId: user._id });
+    let task = await Tasks.Task.findByIdOrAlias(taskId, user._id);
 
-    if (!task) throw new NotFound(res.t('taskNotFound'));
+    if (!task) {
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: same here

File: website/server/controllers/api-v3/tasks.js [hunk 40]
```diff
@@ -634,9 +634,18 @@ api.addTagToTask = {
     if (validationErrors) throw validationErrors;
 
     let taskId = req.params.taskId;
-    let task = await Tasks.Task.findByIdOrAlias(taskId, user._id, { userId: user._id });
+    let task = await Tasks.Task.findByIdOrAlias(taskId, user._id);
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: will this work with tasks that do not have an userId?

File: website/server/controllers/api-v3/tasks.js (hunk 38 above)
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: Hmm actually... this way the leader will be able to edit tags for all tasks linked to the group, even the ones in the users' accounts. Is it okay?

File: website/server/controllers/api-v3/tasks.js (hunk 39 above)
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks.js]: No, I don't believe this is correct. This will now allow a check for challenge or group tasks that do not have a user._id. You can see the previous logic for how this worked with challenges.

File: website/server/controllers/api-v3/tasks.js (hunk 38 above)
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks.js]: I'm not seeing wherein the group.syncTask function that the tags get synced. This should be exactly like challenges. Am I missing something?

File: website/server/controllers/api-v3/tasks.js [hunk 41]
```diff
@@ -312,9 +204,15 @@ api.updateTask = {
 
     let taskId = req.params.taskId;
     let task = await Tasks.Task.findByIdOrAlias(taskId, user._id);
+    let group;
 
     if (!task) {
       throw new NotFound(res.t('taskNotFound'));
+    } else if (task.group.id) {
```
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: @TheHollidayInn continuing from previous discussion because it's not outdated, I see one problem here: you're passing `user._id` to `Tasks.Task.findByIdOrAlias` so a task with no `userId`(like the original ones for challenges and group) will never be found. So lines 217-219 of this file will never be reached.

As for challenges original tasks there should be some special logic for when `task.userId` doesn't exist. If task.userId exists task.challenge or task.groups should not be considered at all except for sanitization (see line 231)

File: website/server/controllers/api-v3/tasks.js [hunk 42]
```diff
@@ -166,7 +166,7 @@ api.getTask = {
 
     if (!task) {
       throw new NotFound(res.t('taskNotFound'));
-    } else if (!task.userId) { // If the task belongs to a challenge make sure the user has rights
+    } else if (task.challenge.id && !task.userId) { // If the task belongs to a challenge make sure the user has rights
```
Comment (khipkin) [inline website/server/controllers/api-v3/tasks.js:169]: Maybe pull this boolean logic into a helper function that takes a task object?  It seems you make this check often enough that it would be beneficial.

Comment (paglias) [reply website/server/controllers/api-v3/tasks.js:169]: I think the check is simple enough that it doesn't require a specific function
